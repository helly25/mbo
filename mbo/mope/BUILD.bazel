# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

load("@bazel_skylib//:bzl_library.bzl", "bzl_library")
load("mope.bzl", "mope_test")

package(default_visibility = ["//visibility:private"])

cc_library(
    name = "ini_cc",
    srcs = ["ini.cc"],
    hdrs = ["ini.h"],
    deps = [
        ":mope_cc",
        "//mbo/file/ini:ini_file_cc",
        "//mbo/status:status_macros_cc",
        "@com_google_absl//absl/container:btree",
        "@com_google_absl//absl/strings",
    ],
    visibility = ["//visibility:public"],
)

cc_library(
    name = "mope_cc",
    srcs = ["mope.cc"],
    hdrs = ["mope.h"],
    deps = [
        "//mbo/types:extend_cc",
        "//mbo/status:status_macros_cc",
        "//mbo/strings:parse_cc",
        "@com_google_absl//absl/cleanup",
        "@com_google_absl//absl/container:node_hash_map",
        "@com_google_absl//absl/log:absl_log",
        "@com_google_absl//absl/functional:function_ref",
        "@com_google_absl//absl/status",
        "@com_google_absl//absl/status:statusor",
        "@com_google_absl//absl/strings",
        "@com_google_absl//absl/strings:str_format",
        "@com_googlesource_code_re2//:re2",
    ],
    visibility = ["//visibility:public"],
)

cc_binary(
    name = "mope",
    srcs = ["mope_main.cc"],
    visibility = ["//visibility:public"],
    deps = [
        ":ini_cc",
        ":mope_cc",
        "//mbo/file:artefact_cc",
        "//mbo/file/ini:ini_file_cc",
        "//mbo/status:status_macros_cc",
        "@com_google_absl//absl/container:btree",
        "@com_google_absl//absl/flags:parse",
        "@com_google_absl//absl/log:initialize",
        "@com_google_absl//absl/status",
        "@com_google_absl//absl/strings",
    ],
)

mope_test(
    name = "mope_test",
    srcs = glob(["tests/*.mope"]),
    outs = glob(["tests/*.golden"]),
    data = glob(["tests/*.ini"]),
)

bzl_library(
    name = "mope_bzl",
    srcs = ["mope.bzl"],
    deps = ["//mbo/diff:diff_bzl"],
    visibility = ["//visibility:public"],
)
