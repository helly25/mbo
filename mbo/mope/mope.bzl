# SPDX-FileCopyrightText: Copyright (c) The helly25/mbo authors (helly25.com)
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Rules mope() and mope(test).

* Rule `mope` takes one or more `srcs` and produces matching `outs`.

* Rule `mope_test` uses `mope` to generate outputs and compares them to existing golden files.
"""

load("@bazel_skylib//lib:dicts.bzl", "dicts")
load("@bazel_skylib//rules:common_settings.bzl", "BuildSettingInfo")
load("//mbo/diff:diff.bzl", "diff_test")

# The `clang-format` tool is selected as follows:
# 1) If custom bazel flag `--//mbo/mope:clang_format` is set, then that value will be used.
# 2) If the custom flag is empty and this variable is a non empty string, then use its value.
#    Otherwise use `@llvm_toolchain_llvm//:bin/clang-format`.
# 3) If the resulting value is `clang-format-auto` (or the above result is not found or not
#    executable), then the rule tries to find the tool:
#    a) `${LLVM_PATH}/bin/clang-format`
#    b) `$(which "clang_format")`
#    c) `clang-format-19` ... `clang-format-14`
#    d) `clang-format` will lastly be picked as a fallback.
CLANG_FORMAT_BINARY = ""  # Ignore clang-format from repo with: "clang-format-auto"

def _get_clang_format(ctx):
    """Get the selected clang-format from `--//mbo/mope:clang_format` bazel flag."""
    return ctx.attr._clang_format_flag[BuildSettingInfo].value

def _clang_format_impl(ctx, src, dst):
    """Clang-format a file.

    Creates an action that performs clang-format on a file and writes the output
    to a separate file.

    Args:
      ctx: The current rule's context object.
      src: The source file.
      dst: The name of the output file.

    Returns:
      The output file.
    """
    clang_config = ctx.files._clang_format_config[0]
    clang_format_tool = [] if CLANG_FORMAT_BINARY else [ctx.executable._clang_format_tool]
    clang_format = _get_clang_format(ctx)
    if not clang_format:
        clang_format = ctx.attr._clang_format_tool if CLANG_FORMAT_BINARY else ctx.executable._clang_format_tool.path
    ctx.actions.run_shell(
        outputs = [dst],
        inputs = [src, clang_config] + clang_format_tool,
        tools = clang_format_tool,
        command = """
            CLANG_FORMAT="{clang_format}"
            if [ "{clang_format}" == "clang-format-auto" ] || [ -x "${{CLANG_FORMAT}}" ]; then
                if [ -x "external/llvm_toolchain_llvm/bin/clang-format" ]; then
                    CLANG_FORMAT="external/llvm_toolchain_llvm/bin/clang-format"
                elif [ -x "${{LLVM_PATH}}/bin/clang-format" ]; then
                    CLANG_FORMAT="${{LLVM_PATH}}/bin/clang-format"
                elif [ $(which "clang_format") ]; then
                    CLANG_FORMAT="clang_format"
                elif [ $(which "clang-format-19") ]; then
                    CLANG_FORMAT="clang-format-19"
                elif [ $(which "clang-format-18") ]; then
                    CLANG_FORMAT="clang-format-18"
                elif [ $(which "clang-format-17") ]; then
                    CLANG_FORMAT="clang-format-17"
                elif [ $(which "clang-format-16") ]; then
                    CLANG_FORMAT="clang-format-16"
                elif [ $(which "clang-format-15") ]; then
                    CLANG_FORMAT="clang-format-15"
                elif [ $(which "clang-format-14") ]; then
                    CLANG_FORMAT="clang-format-14"
                else
                    CLANG_FORMAT="clang-format"
                fi;
            fi;
            # Must cat (<), so that --assume-filename works, so that incldue order gets correct.
            ${{CLANG_FORMAT}} \\
                --assume-filename={assume_filename} \\
                --fallback-style={fallback_style} \\
                --sort-includes={sort_includes} \\
                --style=file:{clang_config} \\
                --Werror \\
                < {src} > {dst} || (echo "CLANG($("${{CLANG_FORMAT}}" --version)) = '${{CLANG_FORMAT}}'" ; false)
            """.format(
            assume_filename = dst.short_path.removesuffix(".gen"),
            clang_format = clang_format,
            clang_config = clang_config.path,
            dst = dst.path,
            fallback_style = ctx.attr._clang_fallback_style,
            sort_includes = "1" if ctx.attr.sort_includes else "0",
            src = src.path,
        ),
        mnemonic = "ClangFormat",
        progress_message = "Clang-Format on file: %s" % (src.path),
        use_default_shell_env = True,
    )

_clang_format_common_attrs = {
    "_clang_fallback_style": attr.string(
        doc = "The fllback stype to pass to clang-format, e.g. 'None' or 'Google'.",
        default = "Google",
    ),
    "_clang_format_flag": attr.label(
        doc = "The flag for the clang-format executable.",
        default = Label("//mbo/mope:clang_format"),
    ),
    "_clang_format_tool": attr.string(
        doc = "The target of the clang-format executable.",
        default = CLANG_FORMAT_BINARY if type(CLANG_FORMAT_BINARY) == "string" else "clang-format",
    ) if CLANG_FORMAT_BINARY else attr.label(
        doc = "The target of the clang-format executable.",
        default = Label("@llvm_toolchain_llvm//:bin/clang-format"),
        allow_single_file = True,
        executable = True,
        cfg = "exec",
    ),
    "_clang_format_config": attr.label(
        doc = "The `.clang-format` file.",
        default = Label("//:clang-format"),
        allow_single_file = [".clang-format"],
    ),
    "sort_includes": attr.bool(
        doc = "Passes --sort-includes to clang-tidy tool if True.",
        default = True,
    ),
}

def _rule_name(str):
    res = ""
    for pos in range(len(str)):
        if str[pos].isalnum():
            res += str[pos]
        else:
            res += "_"
    return res

def _base_no_ext(file):
    # Other files.
    if type(file) == "string":
        name = file.split(":")[-1]
    else:
        name = file.basename
    if not "." in name:
        return name
    return ".".join(name.split(".")[:-1])

def _template_base_no_ext(file):
    # Templates are defined to only accept '.tpl' or '.mope'.
    if type(file) == "string":
        name = file.split(":")[-1]
    else:
        name = file.basename
    if name.endswith(".mope"):
        return name.removesuffix(".mope")
    else:
        return name.removesuffix(".tpl")

def _template_path_no_ext(file):
    # Templates are defined to only accept '.tpl' or '.mope'.
    if type(file) == "string":
        name = file
    else:
        name = file.short_path
    if name.endswith(".mope"):
        return name.removesuffix(".mope")
    else:
        return name.removesuffix(".tpl")

def _gen_name_base_no_ext(file):
    # Remove the optional ".golden" and ".gen" extension and the remaining extension.
    if type(file) == "string":
        name = file.split(":")[-1]
    else:
        name = file.basename
    if name.endswith(".golden"):
        return name.removesuffix(".golden")
    else:
        return name.removesuffix(".gen")

def _build_relative(ctx, file):
    path = "/".join(ctx.build_file_path.split("/")[:-1])
    return file.short_path.removeprefix("../com_helly25_mbo/").removeprefix(path).removeprefix("/")

def _mope_rule_impl(ctx):
    srcs = {_template_base_no_ext(src): src for src in ctx.files.srcs}
    outs = {_gen_name_base_no_ext(out): ctx.actions.declare_file(_build_relative(ctx, out)) for out in ctx.outputs.outs}
    inis = {_base_no_ext(ini): ini.path for ini in ctx.files.data if ini.extension == "ini"}
    if srcs.keys() != outs.keys():
        fail("Files in `srcs` and `outs` do not match on basenames without extensions.")
    for key, src in srcs.items():
        out = outs[key]
        if ctx.attr.clang_format:
            gen = ctx.actions.declare_file(_build_relative(ctx, out) + ".tmp")
        else:
            gen = out
        ctx.actions.run(
            arguments = ctx.attr.args + [
                "--template=" + src.path,
                "--generate=" + gen.path,
                "--ini=" + inis.get(key, ""),
            ],
            executable = ctx.executable._mope_tool,
            inputs = ctx.files.srcs + ctx.files.data,
            mnemonic = "Mope",
            outputs = [gen],
            progress_message = "Mope over " + key,
            tools = ctx.attr._mope_tool.files,
            use_default_shell_env = True,
        )
        if ctx.attr.clang_format:
            _clang_format_impl(ctx, gen, out)
    return [
        DefaultInfo(
            files = depset(ctx.outputs.outs),
            runfiles = ctx.runfiles(files = ctx.outputs.outs),
        ),
    ]

_mope_attrs = {
    "args": attr.string_list(
        doc = "Args to be passed to the mope tool.",
    ),
    "clang_format": attr.bool(
        doc = "Whether to apply clang-format to the generated result.",
        default = False,
    ),
    "data": attr.label_list(
        allow_files = [".ini"],
        default = [],
        doc = "All data files (excl. sources and outs).",
        mandatory = False,
    ),
    "outs": attr.output_list(
        doc = "List of generated file(s). This must match srcs by matching the basenames without extensions (" +
                "e.g. a template `foo.cc.tpl` must have an output `foo.cc`).",
        mandatory = True,
    ),
    "srcs": attr.label_list(
        allow_files = [".tpl", ".mope"],
        doc = "List of templates.",
        mandatory = True,
    ),
    "_mope_tool": attr.label(
        cfg = "exec",
        default = Label("//mbo/mope"),
        doc = "The `mope` tool.",
        executable = True,
    ),
}

mope = rule(
    attrs = dicts.add(
        _mope_attrs,
        _clang_format_common_attrs,
    ),
    doc = "Expand the given `srcs` template(s) to generate `outs`. For each file in `srcs` the required '.tpl' or " +
          "'.mope' extension will be stripped of. The resulting filename must have a corresponding metch in outs " +
          "when a possible '.gen' or '.golden' extension is stripped of.",
    implementation = _mope_rule_impl,
    output_to_genfiles = True,
    provides = [DefaultInfo],
)

def mope_test(
        name,
        srcs,
        outs,
        data = [],
        clang_format = False,
        args = [],
        **kwargs
    ):
    """Run mope over all `srcs` and compare the results with `outs`.

    Args:
        name: Name of the resulting test_suite
        srcs: The mope template(s).
        outs: The corresponding golden generated files.
        data: additional data files (e.g. INI files).ß
        clang_format: Whether to apply clang-fromat to the generated results.
        args: Arguments to pass to the mope tool.
        **kwargs: Other args (e.g. tags, visibility).
    """
    if name.endswith("_test"):
        base_name = name.removesuffix("_test")
    elif name.endswith("_tests"):
        base_name = name.removesuffix("_tests")
    else:
        fail("mope_test rule names must end in '_test' or '_tests'.")
    gens = {_template_base_no_ext(src): _template_path_no_ext(src) + ".gen" for src in srcs}
    outs = {_gen_name_base_no_ext(out): out for out in outs}
    if gens.keys() != outs.keys():
        fail("Files in `srcs` and `outs` do not match on basenames without extensions.")
    mope(
        name = base_name + "_mope",
        args = args,
        clang_format = clang_format,
        data = data,
        outs = gens.values(),
        srcs = srcs,
        testonly = 1,
        **kwargs
    )
    tests = []
    for key, gen in gens.items():
        if len(gens) > 1:
            test_name = base_name + "_" + _rule_name(key) + "_diff_test"
        else:
            test_name = base_name + "_diff_test"
        tests.append(test_name)
        diff_test(
            name = test_name,
            file_old = outs[key],
            file_new = gen,
            failure_message = "The existing old file and the newly generated file differ.",
            **kwargs
        )
    native.test_suite(
        name = name,
        tests = tests,
        **kwargs
    )
