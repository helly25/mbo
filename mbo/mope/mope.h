// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <optional>
#include <string>
#include <string_view>
#include <variant>

#include "absl/container/node_hash_map.h"
#include "absl/functional/function_ref.h"
#include "absl/status/status.h"
#include "mbo/types/extend.h"

namespace mbo::mope {

// MOPE: Mope Over Pump Ends - Is a simple templating system.
//
// Background: Pump.py (Pretty Useful for Meta Programming) is a templating system that allows to expand generic code
// mostly using simple for-loops and conditions. Its drawback is that it is written in Python and that it does not
// support structureed/hierarchical configuration. While moping over possible solutions, the idea came up to implement
// just the necessary dynamic pices combined with a structural templating system.
//
// While more dynamic features might be added in the future, it is expressily not a goal to become turing complete.
// There are many good choices available if that is necessary.
//
// MOPE understands single values and sections which are hierarchical dictionaries that are made up of sections and
// values.
//
// 1) A single value is identified by: '{{' <name> '}}'. The value can be set by calling `SetValue`.
//
// ```c++
// mbo::mope::Templaye mope;
// mope.SetValue("foo", "bar");
// std::string output("My {{foo}}.");
// mope.Expand(output);
// CHECK_EQ(output, "My bar.");
// ```
//
// 2) A section dictionary can be build by calling `AddSubDictionary` multiple times for the same `name` which becomes
// the section name. The section starts with '{{#' <name> '}}' and ends with '{{/' <name> '}};.
//
// ```c++
// mbo::mope::Templaye mope;
// mope.AddSectionDictinary("section")->SetValue("foo", "bar");
// mope.AddSectionDictinary("section")->SetValue("foo", "-");
// mope.AddSectionDictinary("section")->SetValue("foo", "baz");
// mope.AddSectionDictinary("other")->SetValue("many", "more");
// std::string output("My {{#section}}{{foo}}{{/section}}.");
// mope.Expand(output);
// CHECK_EQ(output, "My bar-baz.");
// ```
//
// 3) The template supports for-loops:
//
// '{{# <name> '=' <start> ';' <end> ( ';' <step> )?'}}'...'{{/' <name> '}}'
//
// * <step> is optional and defaults to 1.
// * <step> canot be set to zero.
// * The values <start>, <end> and <step> can either be a number or a name of
//   an existing section dictionary value.
// * <step> > 0: Iteration ends when the current value > <end>.
// * <step> < 0: Iteration ends when the current value < <end>.
//
// This creates an automatic 'section' with a dynamic value under <name> which
// can be accessed by '{{' <name> '}}'.
//
// ```c++
// mbo::mope::Templaye mope;
// mope.SetValue("max", "5");
// std::string output("My {{#foo=1;max;2}}{{foo}}.{{/foor}}");
// mope.Expand(output);
// CHECK_EQ(output, "My 1.3.5.");
// ```
//
// 4) The template allows to set tags from within the template. This allows to
// provide centralized configuration values for instance to for-loops.
//
// These are value tags with a configuration: '{{' <<name> '=' <value> '}}'
//
// ```c++
// mbo::mope::Templaye mope;
// std::string output(R"(
//    {{foo_start=1}}
//    {{foo_end=8}}
//    {{foo_step=2}}
//    My {{#foo=foo_start;foo_end;foo_step}}{{foo}}.{{/foor}})"R);
// mope.Expand(output);
// CHECK_EQ(output, "My 1.3.5.7.");
// ```
class Template {
 public:
  explicit Template() = default;

  // Sets template variable `name` to `value`.
  // The function by default only inserts new values, but if `allow_update` is true, it will also
  // overwrite existing values. It will however fail if an existing value is not a matching value type.
  // Returns whether `name` was newly added (or overwritten).
  absl::Status SetValue(std::string_view name, std::string_view value, bool allow_update = false);

  // Add a sub-dictionary under `name`.
  // The purpose of a sub-dictionary is to be filled, which can be done by calling `SetValue` or `AddSubDictionary`, on
  // the returned value (so it must be used to do so).
  // If the function is called repetedly with the same `name`, then each time a new dictionary for that `name` will be
  // added.
  [[nodiscard]] Template* AddSectionDictionary(std::string_view name);

  // Expands the template `output` in-place.
  [[nodiscard]] absl::Status Expand(std::string& output);

 private:
  enum class TagType {
    kValue,
    kSection,
    kControl,
  };

  struct TagInfo : mbo::types::Extend<TagInfo> {
    std::string name;
    std::string start;
    std::string end;
    std::string config;
    TagType type = TagType::kValue;
  };

  template<class DataType>
  struct TagData : mbo::types::Extend<TagData<DataType>> {
    const TagInfo tag;
    DataType data;
  };

  using SectionDictionary = std::vector<Template>;

  struct Range : mbo::types::Extend<Range> {
    int start = 0;
    int end = 0;
    int step = 1;
    mutable bool expanding = false;
    mutable int curr = 0;
  };

  struct RangeData : mbo::types::Extend<RangeData> {
    std::string start;
    std::string end;
    std::string step;
  };

  // Type `Data` holds all possible information variants. Each of these needs
  // to have a matching `Expand(const TagInfo<Data-Type>&, std::string*)`.
  using Data = std::variant<TagData<SectionDictionary>, TagData<Range>, TagData<std::string>>;

  static std::optional<const Template::TagInfo> FindAndConsumeTag(std::string_view* pos, bool configured_only);

  absl::Status RemoveTags(std::string& output);
  absl::Status MaybeLookup(const TagInfo& tag_info, std::string_view data, int& value) const;
  absl::Status ExpandRangeTag(const TagInfo& tag, Range& range, std::string& output);
  absl::Status ExpandRangeData(const TagInfo& tag, const RangeData& range_data, std::string& output);
  absl::Status ExpandConfiguredTag(const TagInfo& tag, std::string& output);
  absl::Status ExpandTags(
      bool configured_only,
      std::string& output,
      absl::FunctionRef<absl::Status(const TagInfo& tag, std::string&)> func);

  absl::Status Expand(TagData<SectionDictionary>& tag, std::string& output);
  static absl::Status Expand(const TagData<Range>& tag, std::string& output);
  absl::Status Expand(const TagData<std::string>& tag, std::string& output) const;

  template<typename Sink>
  friend void AbslStringify(Sink& sink, const TagType& value);

  absl::node_hash_map<std::string, Data> data_ = {};
};

}  // namespace mbo::mope
