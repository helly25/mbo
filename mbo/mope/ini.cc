// SPDX-FileCopyrightText: Copyright (c) The helly25/mbo authors (helly25.com)
// SPDX-License-Identifier: Apache-2.0
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "mbo/mope/ini.h"

#include <string>
#include <string_view>
#include <utility>
#include <vector>

#include "absl/container/btree_map.h"
#include "absl/status/status.h"
#include "absl/strings/str_split.h"
#include "mbo/file/ini/ini_file.h"
#include "mbo/mope/mope.h"
#include "mbo/status/status_macros.h"

namespace mbo::mope {

absl::Status ReadIniToTemlate(std::string_view ini_filename, Template& root_template) {
  using SectionPath = std::vector<std::pair<std::string, std::string>>;
  MBO_ASSIGN_OR_RETURN(const mbo::file::IniFile ini, mbo::file::IniFile::Read(ini_filename));
  absl::btree_map<SectionPath, mbo::mope::Template*> sections;
  sections[{{"", ""}}] = &root_template;
  for (const std::string& group : ini.GetGroups()) {
    mbo::mope::Template* target = &root_template;
    std::vector<std::string_view> levels = absl::StrSplit(group, '.');
    if (levels.size() == 1 && levels.front().empty()) {
      levels.clear();
    }
    if (!levels.empty()) {
      SectionPath section_path;
      for (const auto& level : levels) {
        std::pair<std::string_view, std::string_view> tail = absl::StrSplit(level, absl::MaxSplits(':', 1));
        section_path.emplace_back(tail.first, tail.second);
        auto [section, inserted] = sections.emplace(section_path, nullptr);
        if (inserted) {
          MBO_ASSIGN_OR_RETURN(section->second, target->AddSection(tail.first));
        }
        target = section->second;
      }
    }
    for (const auto& [key, val] : ini.GetGroupData(group)) {
      MBO_RETURN_IF_ERROR(target->SetValue(key, val));
    }
  }
  return absl::OkStatus();
}

}  // namespace mbo::mope
