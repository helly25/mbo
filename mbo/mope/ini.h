// SPDX-FileCopyrightText: Copyright (c) The helly25 authors (helly25.com)
// SPDX-License-Identifier: Apache-2.0
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#ifndef MBO_MOPE_INI_H_
#define MBO_MOPE_INI_H_
#include <string_view>

#include "absl/status/status.h"
#include "mbo/mope/mope.h"

namespace mbo::mope {

// Reads the INI file `filename` and configures `root_template` with its data.
//
// INI groups are used as sections. They can build a hierarchy:
//
// * The group names are split at '.' to make up the nesting levels.
// * Each level can be repeated by appending a different ':' to the level name.
//
// Example:
//
// ```
// [person]
// id=0
// [person.contact]
// phone=1234
// [person.contact:1]
// phone=2345
// [person:1]
// id=1
// [person:1.contact]
// phone=3456
// [person:1.contact:1]
// phone=4567
// ```
absl::Status ReadIniToTemlate(std::string_view filename, Template& root_template);

}  // namespace mbo::mope

#endif  // MBO_MOPE_INI_H_
