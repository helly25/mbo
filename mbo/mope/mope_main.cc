// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <iostream>
#include <string>
#include <string_view>
#include <utility>
#include <vector>

#include "absl/container/btree_map.h"
#include "absl/flags/flag.h"
#include "absl/flags/parse.h"
#include "absl/flags/usage.h"
#include "absl/log/globals.h"
#include "absl/log/initialize.h"
#include "absl/status/status.h"
#include "absl/strings/str_split.h"
#include "mbo/file/artefact.h"
#include "mbo/file/file.h"
#include "mbo/mope/ini.h"
#include "mbo/mope/mope.h"
#include "mbo/status/status_macros.h"

// NOLINTBEGIN(cppcoreguidelines-avoid-non-const-global-variables,abseil-no-namespace)
ABSL_FLAG(std::string, template, "", "The templte input file (.tpl, .mope).");
ABSL_FLAG(std::string, generate, "-", "The generated out file ('-' for stdout)");
ABSL_FLAG(std::string, ini, "-", "An INI file that can be used to initialize section data.");

ABSL_FLAG(
    std::vector<std::string>,
    set,
    {},
    "A comma separated list of `name=value` pairs, used to seed the template config.");

// NOLINTEND(cppcoreguidelines-avoid-non-const-global-variables,abseil-no-namespace)

namespace mbo {

struct Options {
  std::string template_name;
  std::string generate_name;
};

absl::Status Process(const Options& opts) {
  auto input = mbo::file::Artefact::Read(opts.template_name);
  mbo::mope::Template mope_template;
  // Add `--set` flag values.
  for (const auto& set_kv : absl::GetFlag(FLAGS_set)) {
    const std::pair<std::string_view, std::string_view> kv = absl::StrSplit(set_kv, '=');
    MBO_STATUS_RETURN_IF_ERROR(mope_template.SetValue(kv.first, kv.second));
  }
  // Read `--ini` file if present.
  const std::string ini_filename = absl::GetFlag(FLAGS_ini);
  if (!ini_filename.empty()) {
    MBO_STATUS_RETURN_IF_ERROR(mope::ReadIniToTemlate(ini_filename, mope_template));
  }
  // Expand the template.
  MBO_STATUS_RETURN_IF_ERROR(mope_template.Expand(input->data));
  if (opts.generate_name.empty() || opts.generate_name == "-") {
    std::cout << input->data;
    return absl::OkStatus();
  }
  return mbo::file::SetContents(opts.generate_name, input->data);
}

}  // namespace mbo

static constexpr std::string_view kHelp = R"help(mope --template=<file.mope> [ --generate=<output> ] [ args ]

MOPE: Mope Over Pump Ends - Is a simple templating system.

Args:
  --template=<file.mope>  Path to the mope template.
  --generate=<file.out>   Path to output file. This defaults to '-'- which
                          results in using stdout as output.
  --set=(key=val,)+       List of comma separated `key`, `value` pairs used to
                          set simple values.

Background: Pump.py (Pretty Useful for Meta Programming) is a templating system
that allows to expand generic code mostly using simple for-loops and conditions.
Its drawback is that it is written in Python and that it does not support
structureed/hierarchical configuration. While moping over possible solutions,
the idea came up to implement just the necessary dynamic pices combined with a
structural templating system.

While more dynamic features might be added in the future, it is expressily not a
goal to become turing complete. There are many good choices available if that is
necessary.

MOPE understands single values and sections which are hierarchical dictionaries
that are made up of sections and values.

1) A single value is identified by: '{{' <name> '}}'. The value can be set by
calling `SetValue`.

```c++
mbo::mope::Templaye mope;
mope.SetValue("foo", "bar");
std::string output("My {{foo}}.");
mope.Expand(output);
CHECK_EQ(output, "My bar.");
```

2) A section dictionary can be build by calling `AddSubDictionary` multiple
times for the same `name` which becomes the section name. The section starts
with '{{#' <name> '}}' and ends with '{{/' <name> '}};.

```c++
mbo::mope::Templaye mope;
mope.AddSectionDictinary("section")->SetValue("foo", "bar");
mope.AddSectionDictinary("section")->SetValue("foo", "-");
mope.AddSectionDictinary("section")->SetValue("foo", "baz");
mope.AddSectionDictinary("other")->SetValue("many", "more");
std::string output("My {{#section}}{{foo}}{{/section}}.");
mope.Expand(output);
CHECK_EQ(output, "My bar-baz.");
```

3) The template supports for-loops:

'{{#'<name>'='<start>';'<end>(';'<step>(';'<join> )? )? '}}'...'{{/'<name>'}}'

* The values <start>, <end> and <step> can either be a number or a name of
  an existing section dictionary value.
* <step>:     Is the optional step-difference between iterations and defaults
              to 1. It canot be set to zero.
* <step> > 0: Iteration ends when the current value > <end>.
* <step> < 0: Iteration ends when the current value < <end>.
* <join>:     Optional value that functions as a joiner. The value can be a
              reference or a string in single (') or double quotes (").

This creates an automatic 'section' with a dynamic value under <name> which
can be accessed by '{{' <name> '}}'.

```c++
mbo::mope::Templaye mope;
mope.SetValue("max", "5");
std::string output("My {{#foo=1;max;2;"."}}{{foo}}{{/foor}}");
mope.Expand(output);
CHECK_EQ(output, "My 1.3.5.");
```

4) The template allows to set tags from within the template. This allows to
provide centralized configuration values for instance to for-loops.

These are value tags with a configuration: '{{' <<name> '=' <value> '}}'

```c++
mbo::mope::Templaye mope;
std::string output(R"(
   {{foo_start=1}}
   {{foo_end=8}}
   {{foo_step=2}}
   My {{#foo=foo_start;foo_end;foo_step;'.'}}{{foo}}{{/foor}})"R);
mope.Expand(output);
CHECK_EQ(output, "My 1.3.5.7.");
```

5) The template supports lists:

'{{#' <name> '=[' <values> '] (';' <join>)? }}'...'{{/'<name>'}}'

<values>:     Is a comma separated list which supports (limited, simple only)
              C++ escaping. In addition to the standard C++ escapes, the comma
              ',' curly braces '{', '}' and square braces '[', ']' can also be
              escaped to simplify template writing.
* <join>:     Optional value that functions as a joiner. The value can be a
              reference or a string in single (') or double quotes (").

```c++
mbo::mope::Templaye mope;
std::string output(R"({{#foo=['1,2',3\,4]}}[{{foo}}]{{/foo}})"R);
mope.Expand(output);
CHECK_EQ(output, "[1,2][3,4]");
```

)help";

int main(int argc, char** argv) {
  absl::InitializeLog();
  absl::SetStderrThreshold(absl::LogSeverityAtLeast::kInfo);
  absl::SetProgramUsageMessage(kHelp);
  std::vector<char*> args = absl::ParseCommandLine(argc, argv);
  if (args.size() != 1 || absl::GetFlag(FLAGS_template).empty()) {
    std::cerr << absl::ProgramUsageMessage();
    return 1;
  }
  args.erase(args.begin());
  const auto result = mbo::Process({
      .template_name = absl::GetFlag(FLAGS_template),
      .generate_name = absl::GetFlag(FLAGS_generate),
  });
  if (result.ok()) {
    return 0;
  }
  std::cerr << result << std::endl;
  return 1;
}
