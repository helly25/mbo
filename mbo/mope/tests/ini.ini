subject=Subject

[person]
name=name0
age=25

[person:1]
name=name1
age=33

[person:2]
name=name2
age=42
