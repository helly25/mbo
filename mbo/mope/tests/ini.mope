Subject: {{subject}}

People:
{{#person}}

Name: {{name}}
Age: {{age}}
{{/person}}

Done.
