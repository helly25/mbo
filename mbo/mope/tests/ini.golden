Subject: Subject

People:

Name: name0
Age: 25

Name: name1
Age: 33

Name: name2
Age: 42
